File: benchmark/13_regex_dna/regex_dna.c
// Benchmark 13: Regex DNA Matching
// Category: String Processing / Pattern Matching
// Tests: Pattern matching, string scanning

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <ctype.h>

#define SEQ_LEN 1000000  // 1 million bases

// Simple pattern matching (since we can't use regex library for fair comparison)
// Pattern: Find all occurrences of "ATG" (start codon) followed by any 3 bases, then "TAA", "TAG", or "TGA" (stop codons)

int match_pattern(const char* seq, int len, int pos) {
    // Check for start codon ATG
    if (pos + 11 > len) return 0;
    if (seq[pos] != 'A' || seq[pos+1] != 'T' || seq[pos+2] != 'G') return 0;
    
    // Check for stop codon at position + 6 to + 8 (after 3 any bases)
    char stop[4];
    strncpy(stop, seq + pos + 6, 3);
    stop[3] = '\0';
    
    if (strcmp(stop, "TAA") == 0 || strcmp(stop, "TAG") == 0 || strcmp(stop, "TGA") == 0) {
        return 1;
    }
    return 0;
}

int main() {
    clock_t start = clock();
    
    // Generate random DNA sequence
    char* sequence = (char*)malloc(SEQ_LEN + 1);
    const char* bases = "ATGC";
    
    srand(42);
    for (int i = 0; i < SEQ_LEN; i++) {
        sequence[i] = bases[rand() % 4];
    }
    sequence[SEQ_LEN] = '\0';
    
    // Find pattern matches
    int match_count = 0;
    for (int i = 0; i < SEQ_LEN - 11; i++) {
        if (match_pattern(sequence, SEQ_LEN, i)) {
            match_count++;
        }
    }
    
    clock_t end = clock();
    double time_spent = (double)(end - start) / CLOCKS_PER_SEC;
    
    printf("Sequence length: %d\n", SEQ_LEN);
    printf("Pattern matches found: %d\n", match_count);
    printf("Time: %.4f seconds\n", time_spent);
    
    free(sequence);
    
    return 0;
}

File: benchmark/13_regex_dna/regex_dna.rs
// Benchmark 13: Regex DNA Matching
// Category: String Processing / Pattern Matching
// Tests: Pattern matching, string scanning

use std::time::Instant;

const SEQ_LEN: usize = 1_000_000; // 1 million bases

// Pattern: Find all occurrences of "ATG" (start codon) followed by any 3 bases, then "TAA", "TAG", or "TGA" (stop codons)
fn match_pattern(seq: &[u8], pos: usize) -> bool {
    if pos + 9 > seq.len() {
        return false;
    }

    // Check for start codon ATG
    if seq[pos] != b'A' || seq[pos + 1] != b'T' || seq[pos + 2] != b'G' {
        return false;
    }

    // Check for stop codon at position + 6 to + 8 (after 3 any bases)
    let stop = &seq[pos + 6..pos + 9];
    stop == b"TAA" || stop == b"TAG" || stop == b"TGA"
}

fn main() {
    let start = Instant::now();

    // Generate random DNA sequence
    let bases = [b'A', b'T', b'G', b'C'];
    let mut sequence = Vec::with_capacity(SEQ_LEN);

    let mut rng = Xoroshiro64Star::seed_from_u64(42);
    for _ in 0..SEQ_LEN {
        sequence.push(bases[(rng.next_u64() % 4) as usize]);
    }

    // Find pattern matches
    let match_count = (0..SEQ_LEN - 9)
        .filter(|&i| match_pattern(&sequence, i))
        .count();

    let elapsed = start.elapsed();

    println!("Sequence length: {}", SEQ_LEN);
    println!("Pattern matches found: {}", match_count);
    println!("Time: {:.4} seconds", elapsed.as_secs_f64());
}

// Simple PRNG for reproducibility
struct Xoroshiro64Star {
    state: [u64; 2],
}

impl Xoroshiro64Star {
    fn seed_from_u64(seed: u64) -> Self {
        Self {
            state: [seed, seed.wrapping_add(0x1234567890abcdef)],
        }
    }

    fn next_u64(&mut self) -> u64 {
        let mut s0 = self.state[0];
        let mut s1 = self.state[1];
        let result = s0.wrapping_add(s1).rotate_left(18).wrapping_mul(5);

        s1 ^= s0;
        self.state[0] = s0.rotate_left(24) ^ s1 ^ (s1 << 16);
        self.state[1] = s1.rotate_left(37);

        result
    }
}
